File: apps/restate/src/services/rate-limit.ts
import * as restate from "@restatedev/restate-sdk-cloudflare-workers/fetch";

interface RateLimitState {
  windowStartMs: number;
  count: number;
}

interface RateLimitConfig {
  windowMs: number;
  maxInWindow: number;
}

export const rateLimiter = restate.object({
  name: "RateLimiter",
  handlers: {
    checkAndConsume: async (
      ctx: restate.ObjectContext,
      config: RateLimitConfig,
    ): Promise<void> => {
      const now = await ctx.run("timestamp", () => Date.now());
      const state = (await ctx.get<RateLimitState>("state")) ?? {
        windowStartMs: now,
        count: 0,
      };

      const windowExpired = now - state.windowStartMs >= config.windowMs;
      const current = windowExpired ? { windowStartMs: now, count: 0 } : state;

      if (current.count >= config.maxInWindow) {
        ctx.set("state", current);
        throw new restate.TerminalError("Rate limit exceeded", {
          errorCode: 429,
        });
      }

      ctx.set("state", { ...current, count: current.count + 1 });
    },

    reset: async (ctx: restate.ObjectContext): Promise<void> => {
      ctx.clear("state");
    },
  },
});

export type RateLimiter = typeof rateLimiter;

export function limiter(ctx: restate.Context, key: string) {
  return ctx.objectClient<RateLimiter>(rateLimiter, key);
}

File: apps/restate/src/services/rate_limit.rs
use restate_sdk::prelude::*;
use restate_sdk::serde::Json;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RateLimitConfig {
    pub window_ms: u64,
    pub max_in_window: u64,
}

#[derive(Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
struct RateLimitState {
    window_start_ms: u64,
    count: u64,
}

#[restate_sdk::object]
pub trait RateLimiter {
    #[name = "checkAndConsume"]
    async fn check_and_consume(config: Json<RateLimitConfig>) -> Result<(), HandlerError>;
    async fn reset() -> Result<(), HandlerError>;
}

pub struct RateLimiterImpl;

impl RateLimiter for RateLimiterImpl {
    async fn check_and_consume(
        &self,
        ctx: ObjectContext<'_>,
        config: Json<RateLimitConfig>,
    ) -> Result<(), HandlerError> {
        let config = config.into_inner();
        let now: u64 = ctx
            .run(|| async {
                Ok(std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_millis() as u64)
            })
            .await?;

        let state: RateLimitState = ctx
            .get::<Json<RateLimitState>>("state")
            .await?
            .map(|j| j.into_inner())
            .unwrap_or_default();

        let current = if now - state.window_start_ms >= config.window_ms {
            RateLimitState {
                window_start_ms: now,
                count: 0,
            }
        } else {
            state
        };

        if current.count >= config.max_in_window {
            ctx.set("state", Json(current));
            return Err(TerminalError::new_with_code(429, "rate_limit_exceeded").into());
        }

        ctx.set(
            "state",
            Json(RateLimitState {
                count: current.count + 1,
                ..current
            }),
        );
        Ok(())
    }

    async fn reset(&self, ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
        ctx.clear("state");
        Ok(())
    }
}
